### PR TITLE
Removing more binary_functions

### PR DESCRIPTION
**Pull request by guitargeek**
Follow up to https://github.com/cms-sw/cmssw/pull/24549, with the aim of removing the deprecated `std::binary_function` and `std::unary_function`.

This PR encompasses most of the easier cases, where the inheritance from the unary or binary function could be just removed because it was not necessary.

One other day I will do the "harder" cases in a different PR.

### CODE REVIEW

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by guitargeek** [comment]
Damn I forgot about this new rule...

**Comment by guitargeek** [comment]
Nevermind, it's an actual build error and not a warning. Will fix later

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by guitargeek** [comment]
Ok I was not completely familiar with the new clang warnings yet. I hope fixing the line that the bot pointed out is sufficient.

File: RecoLocalCalo/HcalRecAlgos/src/HBHETimeProfileStatusBitSetter.cc [hunk 1]
```diff
@@ -47,19 +47,21 @@ HBHETimeProfileStatusBitSetter::HBHETimeProfileStatusBitSetter(double R1Min, dou
 HBHETimeProfileStatusBitSetter::~HBHETimeProfileStatusBitSetter(){}
 
 
+bool compareDigiEnergy(const HBHEDataFrame& x, const HBHEDataFrame& y)
```
Comment (slava77) [inline RecoLocalCalo/HcalRecAlgos/src/HBHETimeProfileStatusBitSetter.cc]: I guess ```namespace {``` is in order

File: RecoLocalCalo/HcalRecAlgos/src/HBHETimeProfileStatusBitSetter.cc [hunk 2]
```diff
@@ -47,19 +47,21 @@ HBHETimeProfileStatusBitSetter::HBHETimeProfileStatusBitSetter(double R1Min, dou
...
+{
+  double TotalX=0, TotalY=0;
```
Comment (slava77) [inline RecoLocalCalo/HcalRecAlgos/src/HBHETimeProfileStatusBitSetter.cc]: this edit is a good chance to conform to the cmssw naming style: lower case for variable names

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by guitargeek** [comment]
I didn't expect such a strange failure from this PR. So before I investigate: did you generally make the experience that some unit test randomly can randomly fail?

**Comment by perrotta** [comment]
@guitargeek : the unit test which fails is
``` 
PhysicsTools/PatAlgos/test/patTuple_updateMet_fromMiniAOD_cfg.py
```

It also fails in in the IB since CMSSW_10_4_X_2018-10-12-2300, see 
https://cmssdt.cern.ch/SDT/cgi-bin/showBuildLogs.py/slc7_amd64_gcc700/www/fri/10.4-fri-23/CMSSW_10_4_X_2018-10-12-2300

It apparently succeeds in CMSSW_10_4_X_2018-10-13-1100, even though this IB is identical to the previous one in terms of pull requests, but it restart reporting failure starting from the following CMSSW_10_4_X_2018-10-14-0000

I imagine your PR has nothing to do with it...

**Comment by andrius-k** [comment]
please test

**Comment by andrius-k** [comment]
please test

I accidentally started tests again and then aborted them in a few seconds. Apparently now the bot didn't roll back to `tests-approved` but to `tests-pending` label instead, so running them one more time.

**Comment by andrius-k** [comment]
For some reason tests are still failing. Are you sure everything is ok here?

**Comment by andrius-k** [comment]
@perrotta, ok I see. I thought that tests failed for this reason only once and now they should run properly as my [accidental test run](https://github.com/cms-sw/cmssw/pull/24862#issuecomment-430134324) removed `tests-approved` label. 
It appears that `tests-approved` was added manually, is that correct?

**Comment by slava77** [comment]
@cmsbuild please test

to get cleaner test results
(as well as also having working fwlite-based reco comparisons)

**Comment by andrius-k** [comment]
+1

**Comment by Martin-Grunewald** [comment]
+1

**Comment by fabiocos** [comment]
@rekovic @thomreis @lpernie @tocheng sorry but we need to move forward with this PR, please check and sign it by Monday.

**Comment by thomreis** [comment]
+1

**Comment by fabiocos** [comment]
@tocheng @lpernie @pohsun could you please check and sign it?

**Comment by fabiocos** [comment]
@smuzaffar not sure why the bot did not update the status

**Comment by fabiocos** [comment]
+1

**Comment by fabiocos** [comment]
merge

**Pull request merged**